File: src/concepts.js
'use strict';

// Concept registry (2026-09-06, Grimmethy: "break the existing project down into
// concepts that can each be given this same kind of research treatment... give us a log
// of the work done on these concepts... track how much of the concept we built from
// scratch compared to using resources found in other repos").
//
// A "concept" is a named, narrow topic (e.g. "chat-context-trimming",
// "web-search-capability") that gets the same treatment already run twice by hand this
// session: a research fork surveys other projects, deep-dives them, and files findings
// via writeSideFindingInbox()/side-finding-sweep.js. This module is just the registry
// row + tally counters for that pattern -- the actual audit trail (which brain-dump
// entries and tasks belong to a concept) is a query over existing data
// (getConceptTimeline), not a duplicated log, so it can never drift from the source.
//
// v1 scope is deliberately narrow: concepts are created manually or organically (the
// first research fork on a new topic creates its row), never by an autonomous
// pipeline-driven task source -- see the plan's "explicitly out of scope" section for
// why (mirrors the arch_import premise-check incident's risk shape).

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { writeJsonAtomicSync } = require('./atomic-write.js');

function conceptsPath(pipelineDir) {
  return path.join(pipelineDir, 'concepts.json');
}

// Mirrors apply-group-a.js's loadBrainDump: never let a missing/corrupt file break a
// caller -- an empty registry is always a safe fallback.
function loadConcepts(pipelineDir) {
  const filePath = conceptsPath(pipelineDir);
  let data;
  try {
    data = JSON.parse(fs.existsSync(filePath) ? fs.readFileSync(filePath, 'utf8') : '{"concepts":[]}');
  } catch (err) {
    console.error(`[loadConcepts] ${filePath}: ${err.message} -- returning empty store`);
    data = { concepts: [] };
  }
  if (!Array.isArray(data.concepts)) data.concepts = [];
  return data;
}

function writeConcepts(pipelineDir, data) {
  const filePath = conceptsPath(pipelineDir);
  fs.mkdirSync(path.dirname(filePath), { recursive: true });
  writeJsonAtomicSync(filePath, data);
}

function findConcept(data, id) {
  return data.concepts.find((c) => c && c.id === id) || null;
}

function slugify(name) {
  return String(name || '').toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-+|-+$/g, '').slice(0, 48) || 'concept';
}

// Idempotent on the slugified name -- a second research fork "creating" a concept that
// already exists (organic-creation path, no coordination between callers) must return
// the existing row unchanged, never a duplicate.
// 2026-09-08, Grimmethy: "establish a single, documented task log template ... a blank
// template that can be viewed in the concepts tab" -- every existing concept is a
// narrative research finding (a flat prose `description`, rendered as plain escaped
// text, no markdown). `kind` distinguishes a REFERENCE document (a schema/template, not
// a finding) so the dashboard can render it with real headers/lists instead -- see
// index.html's renderConceptCard. Omitted (undefined) for every existing caller and
// every concept created before this, which is the point: a concept with no `kind` keeps
// today's exact rendering, unaffected.
function createConcept({ name, description, kind }, pipelineDir, { createdBy } = {}) {
  const data = loadConcepts(pipelineDir);
  const slug = slugify(name);
  const existing = data.concepts.find((c) => c && c.slug === slug);
  if (existing) return existing;

  const concept = {
    id: `concept-${slug}-${crypto.randomBytes(3).toString('hex')}`,
    slug,
    name: String(name || '').trim(),
    description: String(description || '').trim(),
    status: 'open',
    createdAt: new Date().toISOString(),
    createdBy: createdBy === 'organic' ? 'organic' : 'manual',
    researchForkCount: 0,
    lastResearchedAt: null,
    builtFromScratchCount: 0,
    adaptedFromResourceCount: 0,
  };
  if (kind === 'reference') concept.kind = 'reference';
  data.concepts.push(concept);
  writeConcepts(pipelineDir, data);
  return concept;
}

// A concept parked (shelved) or finished (shipped) is a deliberate, terminal-for-now
// decision -- mirrors task-disposition.js's own STABLE_TERMINAL_STAGES convention (a
// stable state a sweep/mutation must never silently reopen). Every mutator below that
// would otherwise keep nudging status forward (research, build-tally) now refuses to
// touch a concept in one of these states until it's explicitly reopened -- e.g. a stray
// research fork landing after Task Atomization was shelved for "a day of work, backlog
// takes priority" must not silently flip it back to 'researched' and erase that decision.
const STABLE_CONCEPT_STATUSES = new Set(['shelved', 'shipped']);

function recordConceptResearch(pipelineDir, conceptId) {
  if (!conceptId) return null;
  const data = loadConcepts(pipelineDir);
  const concept = findConcept(data, conceptId);
  if (!concept || STABLE_CONCEPT_STATUSES.has(concept.status)) return null;
  concept.researchForkCount = (concept.researchForkCount || 0) + 1;
  concept.lastResearchedAt = new Date().toISOString();
  if (concept.status === 'open') concept.status = 'researched';
  writeConcepts(pipelineDir, data);
  return concept;
}

// kind: 'scratch' | 'adapted'. Self-reported by the implementing model -- deliberately
// unverified in v1 (see plan's "explicitly out of scope"); the dashboard must label this
// tally as such rather than presenting it as fact.
function recordConceptBuildTally(pipelineDir, conceptId, kind) {
  if (!conceptId || (kind !== 'scratch' && kind !== 'adapted')) return null;
  const data = loadConcepts(pipelineDir);
  const concept = findConcept(data, conceptId);
  if (!concept || STABLE_CONCEPT_STATUSES.has(concept.status)) return null;
  if (kind === 'scratch') concept.builtFromScratchCount = (concept.builtFromScratchCount || 0) + 1;
  else concept.adaptedFromResourceCount = (concept.adaptedFromResourceCount || 0) + 1;
  if (concept.status === 'researched' || concept.status === 'open') concept.status = 'in-progress';
  writeConcepts(pipelineDir, data);
  return concept;
}

// Lifecycle (2026-09-06, Grimmethy: "I envision them as a sort of hub we can return to
// and develop and optimize over time" -- real registry data confirmed the actual gap:
// 14 of 16 tracked concepts were stuck at 'researched' forever, indistinguishable
// between "genuinely forgotten" and "deliberately parked" -- e.g. Task Atomization,
// explicitly scoped as "a job for a whole day of work" and shelved with the backlog
// taking priority, had no way to record that decision at all).
//
// shelveConcept: parks a concept deliberately, with WHY and (optionally) what would make
// it worth reopening -- the two things a future session actually needs to pick this back
// up without re-deriving the same judgment call from scratch. statusBeforeShelve is
// stashed so reopenConcept can restore real progress state (researched vs in-progress)
// instead of resetting every reopened concept back to a blank slate.
function shelveConcept(pipelineDir, conceptId, { reason, revisitCondition } = {}) {
  if (!conceptId || !reason) return null;
  const data = loadConcepts(pipelineDir);
  const concept = findConcept(data, conceptId);
  if (!concept || STABLE_CONCEPT_STATUSES.has(concept.status)) return null;
  concept.statusBeforeShelve = concept.status;
  concept.status = 'shelved';
  concept.shelvedAt = new Date().toISOString();
  concept.shelvedReason = String(reason).trim();
  concept.revisitCondition = revisitCondition ? String(revisitCondition).trim() : null;
  writeConcepts(pipelineDir, data);
  return concept;
}

// reopenConcept: the only way out of 'shelved' (or 'shipped', for the rare "actually,
// more to do here" case) -- restores whatever status the concept held right before it
// was shelved, defaulting to 'researched' for a shipped concept reopened with no prior
// shelve on record. Clears the shelve-specific fields rather than leaving stale ones
// that would misleadingly suggest the concept is still parked.
function reopenConcept(pipelineDir, conceptId) {
  if (!conceptId) return null;
  const data = loadConcepts(pipelineDir);
  const concept = findConcept(data, conceptId);
  if (!concept || !STABLE_CONCEPT_STATUSES.has(concept.status)) return null;
  concept.status = concept.statusBeforeShelve || 'researched';
  delete concept.statusBeforeShelve;
  delete concept.shelvedAt;
  delete concept.shelvedReason;
  delete concept.revisitCondition;
  writeConcepts(pipelineDir, data);
  return concept;
}

// shipConcept: the concept's work is genuinely done, not just "some work happened"
// (which 'in-progress' already means) -- a real terminal state a hub view can use to
// stop surfacing a concept as active work.
function shipConcept(pipelineDir, conceptId) {
  if (!conceptId) return null;
  const data = loadConcepts(pipelineDir);
  const concept = findConcept(data, conceptId);
  if (!concept || STABLE_CONCEPT_STATUSES.has(concept.status)) return null;
  concept.status = 'shipped';
  concept.shippedAt = new Date().toISOString();
  writeConcepts(pipelineDir, data);
  return concept;
}

// Computed on demand from existing data, never persisted -- a concept's audit trail is
// a query over brain-dump.json (findings raised while researching/working on it) and
// task history (implementation work tagged with this concept), so it can't drift from
// the source of truth the way a duplicated log could.
function getConceptTimeline(pipelineDir, conceptId, { loadBrainDump, listTaskHistory } = {}) {
  const rows = [];

  if (typeof loadBrainDump === 'function') {
    let brainDump;
    try {
      brainDump = loadBrainDump(path.join(pipelineDir, 'brain-dump.json'));
    } catch { brainDump = { entries: [] }; }
    for (const entry of (brainDump.entries || [])) {
      if (entry && entry.raisedBy && entry.raisedBy.conceptId === conceptId) {
        rows.push({
          at: entry.capturedAt || entry.lastSeenAt || null,
          kind: 'research-finding',
          ref: entry.id,
          summary: (entry.rawText || '').split('\n')[0].slice(0, 120),
        });
      }
    }
  }

  if (typeof listTaskHistory === 'function') {
    let tasks;
    try {
      tasks = listTaskHistory(pipelineDir) || [];
    } catch { tasks = []; }
    for (const task of tasks) {
      if (task && task.conceptId === conceptId) {
        rows.push({
          at: task.completedAt || task.updatedAt || task.createdAt || null,
          kind: 'task',
          ref: task.id,
          summary: task.title || task.id,
        });
      }
    }
  }

  rows.sort((a, b) => String(a.at || '').localeCompare(String(b.at || '')));
  return rows;
}

// Self-reported build-vs-borrow tally (Part 4). Same house style as
// side-finding.js's SIDE-FINDING: marker (one-line, lenient, strip-and-continue), but
// gated on the CALLER opting a task into it via conceptId -- unlike side-finding, this
// is not injected into every prompt, only ones a human/assistant explicitly tied to a
// concept (see AGENTS.md's "Concept research" section for why this stays manual for v1).
const CONCEPT_BUILD_INSTRUCTION = (
  'This task is tied to tracked concept work. When you finish, end your response with '
  + 'one line: CONCEPT-BUILD: scratch|adapted | <one-sentence detail> -- "scratch" if you '
  + "wrote this from nothing, \"adapted\" if you drew on a specific researched repo or "
  + 'design option, naming which one in the detail. This is a self-report for an audit '
  + 'chart, not graded -- be honest, not impressive.'
);

const CONCEPT_BUILD_MARKER_RE = /^CONCEPT-BUILD:\s*(scratch|adapted)\s*\|\s*(.+)$/im;

function injectConceptBuildInstruction(text) {
  const base = text || '';
  if (base.includes('CONCEPT-BUILD:')) return base;
  return `${base}\n\n${CONCEPT_BUILD_INSTRUCTION}`;
}

// Returns { cleanText, report: {kind, detail} | null }. Lenient: a missing/malformed
// marker just means report is null, never thrown -- the real response must never be
// held hostage by an optional audit tag the model forgot to include.
function extractConceptBuildReport(text) {
  const source = text || '';
  const match = source.match(CONCEPT_BUILD_MARKER_RE);
  if (!match) return { cleanText: source, report: null };
  const kind = match[1].toLowerCase();
  const detail = match[2].trim();
  const cleanText = source.replace(match[0], '').replace(/\n{3,}/g, '\n\n').trim();
  if (!detail) return { cleanText, report: null };
  return { cleanText, report: { kind, detail } };
}

module.exports = {
  conceptsPath,
  loadConcepts,
  writeConcepts,
  findConcept,
  createConcept,
  recordConceptResearch,
  recordConceptBuildTally,
  getConceptTimeline,
  CONCEPT_BUILD_INSTRUCTION,
  injectConceptBuildInstruction,
  extractConceptBuildReport,
  STABLE_CONCEPT_STATUSES,
  shelveConcept,
  reopenConcept,
  shipConcept,
};

File: python/dashboard/routes/concepts.py
import uuid
from datetime import datetime, timezone

from flask import Blueprint, abort, jsonify, request

# NOTE: `GHOST_CONCEPT_ID`, `read_concepts`, `write_concepts` and friends live in app.py,
# which imports this module to register the blueprint -- importing them at module top is
# a circular import that only fails when app.py is the entrypoint (`python app.py`, i.e.
# how the dashboard actually runs), because then `from app import ...` here triggers a
# *second*, re-entrant import of app.py. Pull them in lazily inside each view instead:
# by the time any view runs, app.py is fully initialised and `from app import ...` is
# just a dict lookup. (Same pattern as routes/reports.py.)

concepts_bp = Blueprint("concepts-bp", __name__)


@concepts_bp.route("/api/concepts/<concept_id>/ghost-telemetry")
def api_concept_ghost_telemetry(concept_id):
    """Audit surface for concept-ghost-in-the-machine-0dbeea: how often the pipeline
    recovered a task itself vs. how often a human/agent hand-fixed it (a requeue click).
    404 for any other concept -- this data isn't per-concept, the route is just that
    concept's dashboard hook."""
    from app import GHOST_CONCEPT_ID, _ghost_telemetry
    if concept_id != GHOST_CONCEPT_ID:
        abort(404)
    try:
        days = max(1, min(365, int(request.args.get("days", 30))))
    except (TypeError, ValueError):
        days = 30
    return jsonify({"conceptId": concept_id, **_ghost_telemetry(days)})


@concepts_bp.route("/api/concepts")
def api_concepts():
    """Concept Chart tab's list view -- see AGENTS.md's "Concept research" section for
    what a concept is and why this stays a manually/organically populated registry, not
    an autonomous task source."""
    from app import read_concepts
    return jsonify(read_concepts())


@concepts_bp.route("/api/concepts", methods=["POST"])
def api_concepts_create():
    """Manual concept creation (the dashboard's "+ New Concept" action). Idempotent on
    the slugified name -- matches src/concepts.js's createConcept(), since an organic
    creation from a research fork can race a human creating the same concept by hand."""
    from app import read_concepts, slugify_concept_name, write_concepts
    body = request.get_json(silent=True) or {}
    name = (body.get("name") or "").strip()
    if not name:
        abort(400, description="name is required")
    description = (body.get("description") or "").strip()

    concepts = read_concepts()
    slug = slugify_concept_name(name)
    existing = next((c for c in concepts if c.get("slug") == slug), None)
    if existing:
        return jsonify(existing)

    concept = {
        "id": f"concept-{slug}-{uuid.uuid4().hex[:6]}",
        "slug": slug,
        "name": name,
        "description": description,
        "status": "open",
        "createdAt": datetime.now(timezone.utc).isoformat(),
        "createdBy": "manual",
        "researchForkCount": 0,
        "lastResearchedAt": None,
        "builtFromScratchCount": 0,
        "adaptedFromResourceCount": 0,
    }
    # kind (2026-09-08, Grimmethy: "a blank template that can be viewed in the concepts
    # tab" -- see src/concepts.js's own createConcept() comment): distinguishes a
    # reference/template document from every existing concept's narrative finding, so
    # the frontend can render it with real markdown instead of plain escaped text.
    # Mirrors the Node side's own behavior: omitted entirely unless explicitly
    # 'reference', so an ordinary create (and every existing concept) is unaffected.
    if body.get("kind") == "reference":
        concept["kind"] = "reference"
    concepts.append(concept)
    write_concepts(concepts)
    return jsonify(concept)


@concepts_bp.route("/api/concepts/<concept_id>/shelve", methods=["POST"])
def api_concepts_shelve(concept_id):
    """Parks a concept deliberately -- mirrors src/concepts.js's shelveConcept() exactly
    (this dashboard's concept CRUD is a parallel Python-native implementation, same as
    api_concepts_create above, not a Node subprocess call). Requires a real reason: the
    whole point is capturing WHY a future session shouldn't have to re-derive the same
    judgment call from scratch (2026-09-06, real incident: Task Atomization was scoped
    as "a job for a whole day of work" and shelved for exactly that reason, with no way
    to record it before this)."""
    from app import CONCEPT_STABLE_STATUSES, _find_concept_or_404, read_concepts, write_concepts
    body = request.get_json(silent=True) or {}
    reason = (body.get("reason") or "").strip()
    if not reason:
        abort(400, description="reason is required")
    revisit_condition = (body.get("revisitCondition") or "").strip() or None

    concepts = read_concepts()
    concept = _find_concept_or_404(concepts, concept_id)
    if concept.get("status") in CONCEPT_STABLE_STATUSES:
        abort(409, description=f"concept is already {concept.get('status')} -- reopen it first")

    concept["statusBeforeShelve"] = concept.get("status")
    concept["status"] = "shelved"
    concept["shelvedAt"] = datetime.now(timezone.utc).isoformat()
    concept["shelvedReason"] = reason
    concept["revisitCondition"] = revisit_condition
    write_concepts(concepts)
    return jsonify(concept)


@concepts_bp.route("/api/concepts/<concept_id>/reopen", methods=["POST"])
def api_concepts_reopen(concept_id):
    """The only way out of 'shelved' (or 'shipped', for the rare "actually, more to do
    here" case) -- restores whatever status the concept held right before it was
    shelved, defaulting to 'researched' for a shipped concept reopened with no prior
    shelve on record. Mirrors src/concepts.js's reopenConcept()."""
    from app import CONCEPT_STABLE_STATUSES, _find_concept_or_404, read_concepts, write_concepts
    concepts = read_concepts()
    concept = _find_concept_or_404(concepts, concept_id)
    if concept.get("status") not in CONCEPT_STABLE_STATUSES:
        abort(409, description="concept is not shelved or shipped")

    concept["status"] = concept.pop("statusBeforeShelve", None) or "researched"
    concept.pop("shelvedAt", None)
    concept.pop("shelvedReason", None)
    concept.pop("revisitCondition", None)
    write_concepts(concepts)
    return jsonify(concept)


@concepts_bp.route("/api/concepts/<concept_id>/ship", methods=["POST"])
def api_concepts_ship(concept_id):
    """The concept's work is genuinely done, not just "some work happened" (which
    'in-progress' already means) -- mirrors src/concepts.js's shipConcept()."""
    from app import CONCEPT_STABLE_STATUSES, _find_concept_or_404, read_concepts, write_concepts
    concepts = read_concepts()
    concept = _find_concept_or_404(concepts, concept_id)
    if concept.get("status") in CONCEPT_STABLE_STATUSES:
        abort(409, description=f"concept is already {concept.get('status')}")

    concept["status"] = "shipped"
    concept["shippedAt"] = datetime.now(timezone.utc).isoformat()
    write_concepts(concepts)
    return jsonify(concept)


@concepts_bp.route("/api/concepts/<concept_id>/timeline")
def api_concept_timeline(concept_id):
    """On-demand merge of a concept's research findings (brain-dump entries whose
    raisedBy.conceptId matches) and implementation task history, sorted chronologically --
    a query over existing data, never a duplicated log, so it can't drift from the
    source (see src/concepts.js's getConceptTimeline, this route's Node-side
    equivalent). Task-history coverage is time-budgeted -- see
    _concept_task_history_rows's own header for the 73s live incident this closes --
    so `truncated: true` means the done/-archive history may be incomplete, not that
    nothing was found."""
    from app import _concept_task_history_rows, get_pipeline_dir, read_brain_dump_entries
    pipeline_dir = get_pipeline_dir()
    if not pipeline_dir:
        abort(500, description="no active project configured")

    rows = []
    for entry in read_brain_dump_entries():
        raised_by = entry.get("raisedBy") or {}
        if raised_by.get("conceptId") != concept_id:
            continue
        raw_text = entry.get("rawText") or ""
        rows.append({
            "at": entry.get("capturedAt") or entry.get("lastSeenAt"),
            "kind": "research-finding",
            "ref": entry.get("id"),
            "summary": raw_text.split("\n")[0][:120],
        })
    task_rows, truncated = _concept_task_history_rows(pipeline_dir, concept_id)
    rows.extend(task_rows)
    rows.sort(key=lambda r: r.get("at") or "")
    return jsonify({"rows": rows, "truncated": truncated})
